respy/method_of_simulated_moments.py

```diff
@@ -0,0 +1,140 @@
+"""Estimate models with the method of simulated moments (MSM).
+
+The method of simulated moments is developed by [1], [2], and [3] and an estimation
+technique where the distance between the moments of the actual data and the moments
+implied by the model parameters is minimized.
+
+References
+----------
+
+.. [1] McFadden, D. (1989). A method of simulated moments for estimation of discrete
+       response models without numerical integration. Econometrica: Journal of the
+       Econometric Society, 995-1026.
+.. [2] Lee, B. S., & Ingram, B. F. (1991). Simulation estimation of time-series models.
+       Journal of Econometrics, 47(2-3), 197-205.
+.. [3] Duffie, D., & Singleton, K. (1993). Simulated Moments Estimation of Markov Models
+       of Asset Prices. Econometrica, 61(4), 929-952.
+
+"""
+import functools
+
+import numpy as np
+import scipy as sc
+import pandas as pd
+
+from respy.pre_processing.model_processing import process_params_and_options
+from respy.simulate import get_simulate_func
+
+
+def get_msm_func(
+    params,
+    options,
+    moments,
+    calc_moments,
+    weighting_matrix=None,
+    replacement=0,
+    all_dims=False,
+):
+    """Get the criterion function for estimation with MSM.
+
+    Return a version of the :func:`msm` function where all arguments except the
+    parameter vector are fixed with :func:`functools.partial`. Thus the function can be
+    directly passed into an optimizer or a function for taking numerical derivatives.
+
+    Parameters
+    ----------
+    params : pandas.DataFrame
+        DataFrame containing model parameters.
+    options : dict
+        Dictionary containing model options.
+    moments : np.ndarray
+        Array with shape (n_moments,) containing the moments of the actual data.
+    calc_moments : callable
+        Function to compute the moments from the simulated data.
+    weighting_matrix : np.ndarray
```

Comment:
I think we should support DataFrames and Series here. Most of the time, people scale just by the diagonal of the matrix.